presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -462,116 +366,265 @@ else if (!result.getBlocked().isDone()) {
                 return false;
             }
         }
         return true;
     }
 
-    private List<List<StageExecutionAndScheduler>> getStageExecutions(List<StreamingPlanSection> sections)
+    private Optional<SectionExecution> getLatestSectionExecution(StageId stageId)
     {
-        return sections.stream()
-                .map(section -> stream(forTree(StreamingSubPlan::getChildren).depthFirstPreOrder(section.getPlan())).collect(toImmutableList()))
-                .map(plans -> plans.stream()
-                        .map(StreamingSubPlan::getFragment)
-                        .map(PlanFragment::getId)
-                        .map(this::getStageExecutionInfo)
-                        .collect(toImmutableList()))
-                .collect(toImmutableList());
+        List<SectionExecution> sectionExecutions = this.sectionExecutions.get(stageId);
+        if (sectionExecutions == null || sectionExecutions.isEmpty()) {
+            return Optional.empty();
+        }
+        return Optional.of(getLast(sectionExecutions));
+    }
+
+    private StageId getStageId(PlanFragmentId fragmentId)
+    {
+        return new StageId(session.getQueryId(), fragmentId.getId());
+    }
+
+    private List<SectionExecution> getOrCreateStageExecutions(List<StreamingPlanSection> sections)
+    {
+        ImmutableList.Builder<SectionExecution> result = ImmutableList.builder();
+        for (StreamingPlanSection section : sections) {
+            StageId sectionId = getStageId(section.getPlan().getFragment().getId());
+            List<SectionExecution> executions = sectionExecutions.computeIfAbsent(sectionId, (ignored) -> new CopyOnWriteArrayList<>());
+            if (!queryStateMachine.isDone() && (executions.isEmpty() || getLast(executions).isFailed())) {
+                PlanFragment sectionRootFragment = section.getPlan().getFragment();
+
+                Optional<int[]> bucketToPartition = Optional.empty();
+                OutputBuffers outputBuffers = createDiscardingOutputBuffers();
+                ExchangeLocationsConsumer locationsConsumer = (fragmentId, tasks, noMoreExchangeLocations) -> {};
+
+                if (isRootFragment(sectionRootFragment)) {
+                    bucketToPartition = Optional.of(new int[1]);
+                    outputBuffers = createInitialEmptyOutputBuffers(sectionRootFragment.getPartitioningScheme().getPartitioning().getHandle())
+                            .withBuffer(new OutputBufferId(0), BROADCAST_PARTITION_ID)
+                            .withNoMoreBufferIds();
+                    OutputBufferId rootBufferId = getOnlyElement(outputBuffers.getBuffers().keySet());
+                    locationsConsumer = (fragmentId, tasks, noMoreExchangeLocations) ->
+                            updateQueryOutputLocations(queryStateMachine, rootBufferId, tasks, noMoreExchangeLocations);
+                }
+
+                int attemptId = executions.size();
```

Comment:
It seems that we are only using the last element and the size of this list. Curious if we could only use counter and last element.

Reply:
We use the StageExecutionInfo from previous attempts when we build the StageInfo.  Right now I don't think we do anything with them, but we'll need them for tracking wasted cpu spent on retries, and the total cpu/memory the query ended up using.